ui/e2e/src/pages/DashboardPage.ts

```diff
@@ -62,20 +73,36 @@ export class DashboardPage {
     this.panels = page.getByTestId('panel');
     this.panelHeadings = this.panels.locator('header').getByRole('heading');
 
+    this.variableList = page.getByTestId('variable-list');
+    this.variableListItems = this.variableList.getByTestId('template-variable');
+
     this.panelEditor = page.getByTestId('panel-editor');
+    this.variableEditor = page.getByTestId('variable-editor');
+
+    this.alert = page.getByRole('alert');
   }
 
   async startEditing() {
     await this.editButton.click();
     await this.cancelButton.isVisible();
   }
 
+  async saveChanges() {
+    await this.saveButton.click();
+    await this.editButton.isVisible();
```

Comment:
this might be overkill -- part of me thinks, since this is an assertion that we've exited edit mode, that we could do complete assertions on all 3 buttons:

this.editButton.isVisible()
this.saveButton.isNotVisible()
this.cancelButton.isNotVisible() (not sure what the right syntax is for Playwright)

and we could do the same assertions in `startEditing`

Andddd if we really wanted we could have an assertion function like:

`isInEditMode` or `isEditing` -- and we could use this function in both `startEditing` and `saveChanges`

So the total picture:
```
async isEditing() {
  this.editButton.isVisible()
  this.saveButton.isNotVisible()
  this.cancelButton.isNotVisible()

}

async startEditing() { 
  awat this.editButton.click()
  await this.isEditing()
}

aysnc saveChanges() {
  await this.saveButton.click()
  await !this.isEditing()
  ...

}
```

Reply:
And omg if we really wanted we could check for the dashboard title saying "EDIT ..." too, but that might definitely be overkill :)

Reply:
I'm inclined for the team to iteratively add improvements like that in a follow-up, so I can get the core bit around "we can mutate things in tests" merged, but I love the thoughts around continuing to expand these.